composer/core/state.py

```diff
@@ -540,14 +547,117 @@ def state_dict(self) -> Dict[str, Any]:
 
         return state_dict
 
-    def load_model_state(self, state_dict: Dict[str, Any], strict: bool):
+    def _apply_required_algorithms(
+        self,
+        state: Dict[str, Any],
+        logger: Logger,
+        exclude_algorithms: Optional[List[str]] = None,
+        algorithm_passes: Optional[List[AlgorithmPass]] = None,
+    ):
+        """Applies required algorithms which haven't been specified and aren't in the exclude list.
+
+        Args:
+            state (Dict[str, Any]): State from checkpoint.
+            logger (Logger): Logger to use.
+            exclude_algorithms (List[str], optional): List of algorithm names to exclude. (default: ``None``)
+            algorithm_passes (List[AlgorithmPass], optional): A list of algorithm passes to apply to autoloaded algorithms
+                to sort them into the correct order. (default: ``None``)
+        """
+        import composer.algorithms as algorithms  # type: ignore imports used in `eval(representation)`
+
+        # Get repr of existing algorithms
+        state_algos = {}
+        for algo in self.algorithms:
+            if algo.required_on_load():
+                if type(algo) not in state_algos:
+                    state_algos[type(algo)] = []
+                state_algos[type(algo)].append(algo.__repr__())
+
+        # Gather algorithms to apply
+        missing_algos = set()
+        missing_algo_names = []
+        missing_algo_reprs = []
+        for algo_name, serialized_value in state['algorithms']:
+            if hasattr(algorithms, algo_name) and getattr(algorithms, algo_name).required_on_load():
+                algo = eval(f"algorithms.{serialized_value['repr']}")
+                if exclude_algorithms is None or type(algo).__qualname__ not in exclude_algorithms:
+                    if type(algo) in state_algos and not serialized_value['repr'] in state_algos[type(algo)]:
+                        warnings.warn(
+                            textwrap.dedent(
+                                f"required_on_load algorithm {serialized_value['repr']} was enabled when training the "
+                                f"loaded checkpoint but is now specified in the following forms: {', '.join(state_algos[type(algo)])}."
+                                'Potential parameter discrepancies for this required_on_load algorithm may lead to '
+                                'unexpected behavior, including failing to load weights for some layers.'))
+                    elif type(algo) not in state_algos:
+                        missing_algos.add(algo)
+                        missing_algo_names.append(algo_name)
+                        missing_algo_reprs.append(serialized_value['repr'])
+                        self.algorithms.append(algo)
+
+        # Reorder algorithms based on algorithm_passes from engine
+        algorithms = self.algorithms
```

Comment:
The `algorithms` module defined above: `import composer.algorithms as algorithms` already has this name. Any chance you could rename that or this in order to disambiguate and avoid naming conflicts errors?